tests/ui/uninit.rs

```diff
@@ -20,4 +20,22 @@ fn main() {
 
     // This is OK, because all constitutent types are uninit-compatible.
     let _: (MaybeUninit<usize>, [MaybeUninit<bool>; 2]) = unsafe { MaybeUninit::uninit().assume_init() };
+
+    let mut x: MaybeUninit<usize> = MaybeUninit::uninit();
+    let _: usize = unsafe { x.assume_init() };
+
+    let mut x: MaybeUninit<[u8; 0]> = MaybeUninit::uninit();
+    let _: [u8; 0] = unsafe { x.assume_init() };
+
+    let mut x: MaybeUninit<()> = MaybeUninit::uninit();
+    let _: () = unsafe { x.assume_init() };
+
+    let x: MaybeUninit<MaybeUninit<usize>> = MaybeUninit::uninit();
+    let _: MaybeUninit<usize> = unsafe { x.assume_init() };
+
+    let x: MaybeUninit<(MaybeUninit<usize>, MaybeUninit<bool>)> = MaybeUninit::uninit();
+    let _: (MaybeUninit<usize>, MaybeUninit<bool>) = unsafe { x.assume_init() };
+
+    let x: MaybeUninit<(MaybeUninit<usize>, [MaybeUninit<bool>; 2])> = MaybeUninit::uninit();
+    let _: (MaybeUninit<usize>, [MaybeUninit<bool>; 2]) = unsafe { x.assume_init() };
```

Comment:
I'd like to see a false-positive check here, e.g.

```
let x: MaybeUninit<usize> = MaybeUninit::uninit();
let _ = unsafe {
    x.as_mut_ptr().write(42usize);
    x.assume_init()
}
```

Reply:
I marked it as WIP as I was just thinking about that. Will need to detect cases where the `MaybeUninit` variable is initialized before a call to `assume_init` and ignore that case.